build.rs

```diff
@@ -170,147 +204,149 @@ fn probe_r_paths() -> io::Result<InstallationPaths> {
+    let r_version_split = r_version
+        .split('.')
+        .map(|s| {
+            // Good:
+            //   - "4.1.2"
+            //
+            // Bad:
+            //   - "4.1.foo" (some part contains any non-digit characters)
+            //   - "4.1." (some part is missing)
+            if !s.is_empty() && s.chars().all(|c| c.is_digit(10)) {
+                Some(s)
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<Option<&str>>>();
+
+    let (major, minor, patch) = match *r_version_split.as_slice() {
+        // if any of the first three items doesn't exist, the format is invalid
+        [] | [None, ..] => return Err(EnvVarError::InvalidEnvVar("Cannot find R major version")),
+        [_, None, ..] => return Err(EnvVarError::InvalidEnvVar("Cannot find R minor version")),
+        [_, _, None, ..] => return Err(EnvVarError::InvalidEnvVar("Cannot find R patch level")),
+        // if all of the first three items exist, the format is valid
+        [Some(major), Some(minor), Some(patch)] => {
+            (major.to_string(), minor.to_string(), patch.to_string())
+        }
+        // if the length is longer than 3, the format is invalid
+        _ => return Err(EnvVarError::InvalidEnvVar("Invalid format")),
+    };
 
-    if input.len() == 4 {
-        if input[3] == "devel" {
-            result.devel = "-devel".into();
-        } else {
-            return Err(EnvVarError::InvalidEnvVar("Cannot find R development status"));
+    let devel = match devel {
+        Some("devel") => true,
+        Some(_) => {
+            return Err(EnvVarError::InvalidEnvVar(
+                "Cannot find R development status",
+            ))
         }
-    }
+        None => false,
+    };
 
-    Ok(result)
+    Ok(RVersionInfo {
+        major,
+        minor,
+        patch,
+        devel,
+        version_string,
+    })
 }
 
-
-fn get_r_version_from_env(r_version_env_var:&str) -> Result<RVersionInfo, EnvVarError> {
+fn get_r_version_from_env(r_version_env_var: &str) -> Result<RVersionInfo, EnvVarError> {
     std::env::var(r_version_env_var)
         // Any error arising from reading env var is converted to this value
         .map_err(|_| EnvVarError::EnvVarNotPresent)
-        .map(|v| {
-            v.split(&['.', '-'][..])
-                .map(|s| s.to_string())
-                .collect::<Vec<String>>()
-        })
-        .and_then(str_vec_to_version)
+        .and_then(|s| parse_r_version(s, None))
 }
 
 fn get_r_version_from_r(r_paths: &InstallationPaths) -> Result<RVersionInfo, EnvVarError> {
-    let r_binary = if cfg!(target_os = "windows") {
-        Path::new(&r_paths.library)
-            .join("R.exe")
+    let r_binary = if cfg!(windows) {
+        Path::new(&r_paths.library).join("R.exe")
     } else {
-        Path::new(&r_paths.r_home)
-            .join("bin")
-            .join("R")
+        Path::new(&r_paths.r_home).join("bin").join("R")
```

Comment:
I feel like this also repeats several times, perhaps a `fn get_r_binary(r_paths: &InstallationPaths) -> Path` would be better?

Reply:
Sounds good, thanks.